### PR TITLE
Pass for specialization of shape-relevant inputs

### PR DESCRIPTION
**Pull request by aprocter**
This adds a pass for compile-time specialization (fancy word for selective constant-folding) of shape-relevant inputs. For example, the `shape` input of the forthcoming `DynReshape` will need to be evaluated at compile time (or at least at parameter-binding time) in order to be converted to a static `Reshape`.

There are a couple of backwards-compatible changes to `Node` here:

(1) a new `as_constants` virtual that can be used to replace a node with a sequence list of `op::Constant` nodes (assuming its inputs are themselves all constants). Default behavior is for conversion to fail [which users must be able to deal with], and overriding this is optional---so as_constants needs to be sound if implemented, but not complete.

(2) new facilities to tag inputs as `shape-relevant` (meaning that their value has an effect on the output shapes of the node) or as `value-irrelevant` (meaning that their value does *not* have an effect on the output *values* of the node; AFAIK the only example of this at the moment is `ShapeOf`). By default, inputs are assumed to be shape-irrelevant and value-relevant, so it's only the new fancy dynamic ops that will need to mess with any of this.

**EDIT**: One more thing I slipped in here: `ShapeOf` now returns `i64`, not `u64`. Since `ShapeOf` is an experimental op, I don't think this warrants a `Changes API` tag.

### CODE REVIEW

File: src/ngraph/descriptor/input.hpp [hunk 1]
```diff
@@ -82,8 +92,12 @@ namespace ngraph
             Node* m_node;   // The node we are an input for
             size_t m_index; // Index into all input tensors
             Output* m_output;
+            bool m_is_relevant_to_shape;
```
[the remark this answers is not in the document]
Comment (aprocter) [reply src/ngraph/descriptor/input.hpp]: These are actually accessed directly from `Node`,  which is a `friend`. I think I'll resolve by taking the variables private and removing the mutators.

File: src/ngraph/node.hpp [hunk 2]
```diff
@@ -109,6 +110,22 @@ namespace ngraph
         // Called after transition
         void delayed_validate_and_infer_types();
 
+        /// \brief Produce a vector of constant nodes (one for each of this node's outputs) that
+        ///        can replace this node's outputs. May return an empty vector to signal that
+        ///        conversion to constants is not possible or not supported.
+        /// \returns If conversion is successful, a vector of op::Constant nodes, corresponding
+        ///          to this node's outputs in order. If unsuccessful, an empty vector.
+        ///
+        /// Conversion does not have to be complete. That means that subclasses *may* override
```
[the remark this answers is not in the document]
Comment (aprocter) [reply src/ngraph/node.hpp]: I don't think that's true---the "dangling" paragraph is a standard way to supply a detailed description.

File: src/ngraph/pass/shape_specialization.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/pass/shape_specialization.hpp"
+#include "ngraph/graph_util.hpp"
+#include "ngraph/op/constant.hpp"
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (aprocter) [reply src/ngraph/pass/shape_specialization.cpp]: I'll just remove the `using` here

File: src/ngraph/op/concat.cpp [hunk 4]
```diff
@@ -93,6 +94,48 @@ shared_ptr<Node> op::Concat::copy_with_new_args(const NodeVector& new_args) cons
...
+    if (get_concatenation_axis() != 0)
+    {
+        return {};
+    }
+
+    size_t total_elements = 0;
+
+    for (size_t i = 0; i < get_input_size(); i++)
+    {
+        if (get_input_element_type(i) != element::i64)
```
[the remark this answers is not in the document]
Comment (aprocter) [reply src/ngraph/op/concat.cpp]: Added a comment a few lines up.

File: src/ngraph/descriptor/input.hpp [hunk 5]
```diff
@@ -84,6 +94,8 @@ namespace ngraph
             Output* m_output;
 
         private:
+            bool m_is_relevant_to_shape;
```
Comment (diyessi) [inline src/ngraph/descriptor/input.hpp]: These might want to be sets of outputs (or indices) and we might want outputs to know sets of inputs they depend on. We'll see. Fine for now.

File: src/ngraph/op/concat.cpp [hunk 6]
```diff
@@ -93,6 +94,52 @@ shared_ptr<Node> op::Concat::copy_with_new_args(const NodeVector& new_args) cons
...
+    }
+
+    std::vector<int64_t> values(total_elements);
+
+    size_t pos = 0;
+
+    for (size_t i = 0; i < get_input_size(); i++)
+    {
+        auto const_node = static_pointer_cast<op::Constant>(get_argument(i));
+        memcpy(values.data() + pos,
```
Comment (diyessi) [inline src/ngraph/op/concat.cpp]: Klocwork will probably flag this `memcpy` so please add a comment justifying why it's safe.

Comment (aprocter) [reply src/ngraph/op/concat.cpp]: Good call---I actually added an `assert` to be extra sure that the safety argument holds. :)

**Pull request merged**